how-we-work/how-to-upgrade-dependencies.md

```diff
@@ -0,0 +1,75 @@
+
+Finally, besides the `pip` installed packages, please also take a look at the Dockerfile (in the top level directory `cellxgene/Dockerfile`) to ensure that there are no security vulnerabilities there. Snyk is your friend to figure out if there are and what the remediation process is.
+
+### Upgrading cellxgene Explorer desktop backend
+
+The policy we have to upgrade the backend of cellxgene Explorer _desktop_ is different than any of the other policies. Because many users may have different versions of the dependent `pip` packages already installed on their laptop, and because many users do not use a virtual environment, we need to be _as permissive as possible_ when defining what the allowable versions of each package are.
+
+To update the `pip` packages for cellxgene Explorer desktop backend, we recommend doing the following steps:
+
+1. Install [safety](https://pypi.org/project/safety/): `pip install safety`. Note: this will ONLY check for vulnerabilities against the currently installed packages.
+
+2. Manually install the minimum required version of all packages in `cellxgene/backend/server/requirements.txt` and `cellxgene/backend/server/requirements-prepare.txt`.
+
+3. Run `safety`: `python -m safety check --full-report`
+
+4. Update requirements as indicated, goto step 2 again.  Repeat until clean.
+
+5. Check that cellxgene still works :)
+
+Please also upgrade the requirements in `cellxgene/backend/server/requirements-dev.txt` though the policy on being as permissive as possible does not apply since only cellxgene Explorer developers need to install those requirements.
+
+[Here](https://github.com/chanzuckerberg/cellxgene/pull/2172/files) is an example PR.
+
+## Upgrade cellxgene Portal
+
+### Upgrading cellxgene Portal frontend
+
+The instructions to upgrade the Portal frontend are exactly the same as the [Explorer frontend instructions](#Upgrading-cellxgene-Explorer-frontend). The only difference is that the package location is `corpora-data-portal/frontend/package.json`.
+
+You will also need to ensure that the frontend Docker image is up to date at `corpora-data-portal/frontend/Dockerfile`.
+
+### Upgrading cellxgene Portal backend
+
+The instructions to upgrade the Portal backend are pretty much identical to the [Explorer CZI-hosted backend instructions](#Upgrading-cellxgene-Explorer-czi-hosted-backend). The following is the list of files to check:
+
+- `corpora-data-portal/requirements.txt`
+- `corpora-data-portl/backend/chalice/api_server/requirements.txt`
+- `corpora-data-portal/backend/chalice/cloudfront_invalidator/requirements.txt`
+- `corpora-data-portal/backend/chalice/upload_failures/requirements.txt`
+- `corpora-data-portal/Dockerfile`
+- `corpora-data-portal/Dockerfile.processing_image`
+- `corpora-data-portal/Dockerfile.upload_failures`
```

Comment:
```suggestion
- `corpora-data-portal/.happy/requirements.txt`
- `corpora-data-portal/Dockerfile.upload_failures`
```

Reply:
Do we not use the Dockerfile.upload_failures at all?

Reply:
Sorry was trying to add the happy requirements. Didnt mean to remove upload failures